### ISSUE TITLE
EBUSY Error when doing npm install/prune/update

### ISSUE THREAD

**Issue by KMayne** [npm/npm#18425]
#### I'm opening this issue because:

  - [x] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
When I install anything locally to my project I get an EBUSY error for a .nfs* file in node_modules/fsevents/node_modules/{package}. 
{package} is typically either abbrev, aproba, attrib.
The fsevents package isn't even installed as I'm on Linux.

Update:
Can reproduce on Windows, getting an EPERM error, presumably for the same reasons.

##### Things I've tried
- Deleting everything: ~/.npm; node_modules; reinstalling npm@5.4.0.
- npm canary (still get the issue)
- Using npm 3.5.2 - no longer had the issue

EDIT:
I observed the following:
With a fresh install of my dependencies, there is no `fsevents` in my `node_modules` directory.
After attempting to install `nodemon` for instance (depends on `fsevents` optionally & transitively), `fsevents` was put into my `node_modules` directory. When I used npm 3 to install the package, npm 3 cleared these up. `fsevents` only shows up in the `node_modules` directory i.e. it's not in the output of `npm list` or in `package-lock.json`.

The `.nfs*` files that seem to be causing the errors look like the `package.json` files of the dependencies of `fsevents`.

The process holding the file descriptor is npm as show in this output of lsof.

| COMMAND | PID | USER | FD | TYPE | DEVICE | SIZE/OFF | NODE | NAME |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
|npm | 1249449 | km3315  | 32w  | REG  | 0,59  | 1412  | [removed] | node_modules/fsevents/node_modules/aproba/.nfs00000002c156c00300003e19 |

I don't know how optional dependencies work in npm but it looks like the problem might be optional dependencies being put in the directory then deleted before some other part of npm has finished with it.

#### How can the CLI team reproduce the problem?
~~Unknown how to reproduce. My colleague working on the same project does not get the error but the post-install stage does take a long time for her.~~
EDIT 2:
1. Start with initial project directory (i.e. `npm init` has been done)
2. Install a package that doesn't depend on fsevents (including transitively) e.g. object.values
3. Attempt to install nodemon
4. Observe error

~~The directory is stored on a network mounted disk.~~
Update: I have tested on Windows on a local disk and get the same issue.
I suspect this is related to installing a package that depends (optionally) on something e.g. fsevents on a non-macOS system.

https://gist.github.com/KMayne/67a5b00ec091dca0af0b2b895fc2d2f7

### supporting information:

 - `npm -v` prints: 5.4.0
 - `node -v` prints: v6.11.1
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Linux & Windows
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by KMayne** [comment]
Update:
When running with --verbose, I've noticed a load of the following lines in the log:
```
npm info lifecycle undefined~preuninstall: undefined
npm info lifecycle undefined~uninstall: undefined
```

I suspect this is related as it looks like the same number of directories in `node_modules/fsevents/node_modules`

Full output with --verbose: https://gist.github.com/KMayne/149abb22b6b4fdd9c452c2ef075be186

Note the above is with the errant fsevents dir

**Comment by KMayne** [comment]
Closing as duplicate of https://github.com/npm/npm/issues/18380

**Comment by KMayne** [comment]
Reopening as bug has been fixed on Windows but not Linux

**Comment by KMayne** [comment]
I can confirm that setting fsync to true as in this commit solves the issue. https://github.com/npm/npm/commit/0b28ac72d29132e9b761717aba20506854465865

**Comment by KMayne** [comment]
However I suspect the root cause may be related to this
https://github.com/npm/npm/issues/18380#issuecomment-326954754

> I've done some git bisecting on on the npm repo and the bug is introduced in this commit: f4650b5

> This updates write-file-atomic to 2.3.0. I've confirmed this by taking the 5.4.0 code and manualy installing write-file-atomic at the previous version (2.1.0) and it fixes the issue.

**Comment by marcins** [comment]
@KMayne: I believe `write-file-atomic` was rolled back to 2.1.0 in npm 5.4.2: https://github.com/npm/npm/releases/tag/v5.4.2

Have you tried npm 5.4.2?

**Comment by KMayne** [comment]
Yep this is fixed in 5.4.2
